RFC-0003-arrow-flight-connector.md

```diff
@@ -0,0 +1,70 @@
+# **RFC0003 for Presto**
+
+## Arrow Flight connector template in Presto
+
+Proposers
+
+* Elbin Pallimalil
+
+## [Related Issues]
+
+## Summary
+
+Create a base Arrow Flight module in Presto which can be extended to create an Arrow Flight connector.
+
+## Background
+
+Apache Arrow Flight is a new general-purpose client-server framework that simplifies high performance transport of large datasets over network interfaces.
+
+Presto currently supports different datasources through connectors for each type of datasource. The connectors could be using JDBC or some other protocol for datasource connection. If we use Flight client to connect to the different datasources we can improve the performance of the datasource connection. Also we would only need to develop one Flight client that can be used across different datasources instead of developing individual connectors for each type of datasource.
+
+## Proposed Implementation
+
+A template for Flight connector will be created as `presto-base-arrow-flight` module. Since each implementation of flight server will be different, this base module can be extended to create a flight connector that supports a particular implementation.
+
+### High level design
+
+#### Proposed design
+
+![Design diagram](arrow-flight-connector/Presto-flight-client.drawio.png)
+
+The Arrow Flight libraries provide a development framework for implementing a service that can discover metadata and send and receive data streams. Flight APIs are used to retrieve the data from the corresponding table selected. The steps below need to be followed. 
+
+1. Create a Flight Client using the host, port, ssl information of the Flight service where the route is opened. 
+2. Use the Flight Client to authenticate to the Flight service. 
+3. Create a FlightDescriptor using a JSON document containing the description of the connection to read or write. The description consists of: 
+    1. Database name, host, port, username and password 
+    2. Interaction properties including schema and table selected or the select statement 
+4. Use the FlightDescriptor to obtain a FlightInfo object. 
+5. Use the FlightInfo object to obtain the list of endpoints we will be retrieving data from. 
+6. Use each endpoint to obtain a Ticket. 
+7. Use the Ticket to obtain a Stream. 
+8. Use the Stream to read data. 
+
+#### Query execution using flight
+
+![Query execution using Flight](arrow-flight-connector/Query-execution-using-flight.png)
+
+### Low level design
+
+#### Connector splits in Presto for FlightInfo tickets
+
+Flight server can serve data using multiple tickets based on the connector type and query. Data can be fetched in parallel using the multiple flight tickets. Presto can create connector splits for each Flight Info ticket so that data can be fetched in parallel by different worker nodes.
+
+![Connector splits in Presto](arrow-flight-connector/Connector-splits-in-Presto.png)
+
+#### Class diagram
```

Comment:
ArrowQueryBuilder seems to reference SQL.  Why is this?  Is there any presumption regarding the format of the Arrow server in this template?

Reply:
We have tested with a Flight server implementation that accepts SQL statements for read operations. That is why we have a reference to query builder.

Reply:
@tdcmeehan The proposed implementation is such that only read operations are supported. And for read operations we are building the query and sending this along with table and schema details to the concrete implementation to create a FlightDescriptor. Do you want us to remove any reference to the query from this connector template?

Reply:
Yes--the implementation must not presume that the underlying flight service accepts SQL.  Please remove any reference to queries and SQL>